marmelab/react-admin
# Introduce `<FilterLiveForm>`

## Problem

The `FilterList` is designed for buttons - not inputs. We provide the `FilterLiveSearch` component, which is a `TextInput` inside a form that submits on change. But some users want to be able to add custom form inputs (e.g. to select an author from a resource using an autocomplete), and re-writing the form wrapper is cumbersome.

## Solution

Introduce a new component to wrap form inputs in that case:

```tsx
export const EventFilterAside = () => (
    <Card>
        <CardContent>
            <FilterList label="Has ordered" icon={<MonetizationOnIcon />}>
                <FilterListItem
                    label="True"
                    value={{
                        nb_commands_gte: 1,
                        nb_commands_lte: undefined,
                    }}
                />
                <FilterListItem
                    label="False"
                    value={{
                        nb_commands_gte: undefined,
                        nb_commands_lte: 0,
                    }}
                />
            </FilterList>
            <FilterLiveForm>
                <TextInput source="q" />
            </FilterLiveForm>
            <FilterLiveForm>
                <ReferenceInput source="authorId" reference="authors" />
            </FilterLiveForm>
        </CardContent>
    </Card>
);
```

This PR also fixes #5591 :tada: 

## Todo

- [x] Introduce `<AutoSubmitFilterForm>`
- [x] Introduce `<FilterListWrapper>` to make it look nicer next to `<FilterList>`
- [x] Add stories
- [x] Write doc
- [x] Try to fix [debounce issue](https://github.com/marmelab/react-admin/issues/5591)
  - [x] Story
  - [x] Check if the behavior when there is validation at play is OK
- [x] Try to fix issue where `FilterLiveSearch` keeps applying old value
- [x] Use `AutoSubmitFilterForm` in `FilterForm` to keep the code more consistent
- [x] Use `AutoSubmitFilterForm` in `FilterLiveSearch` to keep the code more consistent

## How To Test

- `<FilterLiveForm>` logic: http://localhost:9010/?path=/story/ra-core-controller-list-filter-filterliveform--basic
- `<FilterLiveForm>` UI: http://localhost:9010/?path=/story/ra-ui-materialui-list-filter-filterliveform--full-app
- `<FilterListSection>`: http://localhost:9010/?path=/story/ra-ui-materialui-list-filter-filterlistsection--basic
- Filter debounce bug: http://localhost:9010/?path=/story/ra-ui-materialui-input-searchinput--race-condition
- `<FilterLiveSearch>` old value bug: http://localhost:9010/?path=/story/ra-ui-materialui-list-filter-filterlivesearch--full-app
  1. Enter a filter value (e.g. 'foo')
  2. Navigate to Countries
  3. Navigate back to Books
  4. Clear the `<FilterLiveSearch>`
  5. Confirm the old value ('foo') does not come back

## Additional Checks


Also, please make sure to read the [contributing guidelines](https://github.com/marmelab/react-admin#contributing).
